File: CashuWallet/Core/Protocols/StorageProtocol.swift
import Foundation

// MARK: - Storage Protocol

/// Protocol for persistent storage operations.
/// Abstracts storage implementation to allow for different backends (UserDefaults, SQLite, Keychain, etc.)
protocol StorageProtocol {
    /// Store a value for a key
    func set<T: Codable>(_ value: T, forKey key: String) throws
    
    /// Retrieve a value for a key
    func get<T: Codable>(forKey key: String) throws -> T?
    
    /// Remove a value for a key
    func remove(forKey key: String) throws
    
    /// Check if a key exists
    func exists(forKey key: String) -> Bool
    
    /// Get all keys with a given prefix
    func keys(withPrefix prefix: String) -> [String]
}

// MARK: - Secure Storage Protocol

/// Protocol for secure storage (Keychain)
protocol SecureStorageProtocol {
    /// Store a secret securely
    func saveSecret(_ secret: String, forKey key: String) throws
    
    /// Retrieve a secret
    func loadSecret(forKey key: String) throws -> String?
    
    /// Delete a secret
    func deleteSecret(forKey key: String) throws
    
    /// Check if a secret exists
    func hasSecret(forKey key: String) -> Bool
}

// MARK: - UserDefaults Storage Implementation

/// Storage implementation using UserDefaults
final class UserDefaultsStorage: StorageProtocol {
    private let defaults: UserDefaults
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()
    
    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }
    
    func set<T: Codable>(_ value: T, forKey key: String) throws {
        let data = try encoder.encode(value)
        defaults.set(data, forKey: key)
    }
    
    func get<T: Codable>(forKey key: String) throws -> T? {
        if let data = defaults.data(forKey: key) {
            return try decoder.decode(T.self, from: data)
        }

        // Legacy compatibility for values previously written directly to UserDefaults.
        return defaults.object(forKey: key) as? T
    }
    
    func remove(forKey key: String) throws {
        defaults.removeObject(forKey: key)
    }
    
    func exists(forKey key: String) -> Bool {
        defaults.object(forKey: key) != nil
    }
    
    func keys(withPrefix prefix: String) -> [String] {
        defaults.dictionaryRepresentation().keys.filter { $0.hasPrefix(prefix) }
    }
}

// MARK: - Storage Keys

/// Centralized storage key definitions
enum StorageKeys {
    static let walletDataPrefix = "wallet."
    static let npcDataPrefix = "npc."

    // Wallet
    static let mints = "wallet.mints"
    static let activeMintUrl = "wallet.activeMintUrl"
    static let pendingTokens = "wallet.pendingTokens"
    static let pendingReceiveTokens = "wallet.pendingReceiveTokens"
    static let claimedTokens = "wallet.claimedTokens"
    static let transactions = "wallet.transactions"
    static let savedTokens = "wallet.savedTokens"
    static let paymentPreimages = "wallet.paymentPreimages"
    static let meltQuoteFees = "wallet.meltQuoteFees"
    static let mintQuoteTimestamps = "wallet.mintQuoteTimestamps"
    static let processedNPCQuotes = "wallet.processedNPCQuotes"
    
    // Settings
    static let useBitcoinSymbol = "settings.useBitcoinSymbol"
    static let showFiatBalance = "settings.showFiatBalance"
    static let bitcoinPriceCurrency = "settings.bitcoinPriceCurrency"
    static let checkPendingOnStartup = "settings.checkPendingOnStartup"
    static let checkSentTokens = "settings.checkSentTokens"
    static let autoPasteEcashReceive = "settings.autoPasteEcashReceive"
    static let useWebsockets = "settings.useWebsockets"
    static let enablePaymentRequests = "settings.enablePaymentRequests"
    static let receivePaymentRequestsAutomatically = "settings.receivePaymentRequestsAutomatically"
    static let enableNWC = "settings.enableNWC"
    static let nwcConnections = "settings.nwcConnections"
    static let showP2PKButtonInDrawer = "settings.showP2PKButtonInDrawer"
    static let p2pkKeys = "settings.p2pkKeys"
    static let checkIncomingInvoices = "settings.checkIncomingInvoices"
    static let periodicallyCheckIncomingInvoices = "settings.periodicallyCheckIncomingInvoices"
    static let nostrRelays = "settings.nostrRelays"
    static let nostrSignerType = "settings.nostrSignerType"
    static let amountDisplayPrimary = "settings.amountDisplayPrimary"

    enum Legacy {
        static let mints = "savedMints"
        static let pendingTokens = "pendingTokens"
        static let pendingReceiveTokens = "pendingReceiveTokens"
        static let claimedTokens = "claimedTokens"
        static let savedTokens = "savedTokens"
        static let paymentPreimages = "paymentPreimages"
        static let mintQuoteTimestamps = "mintQuoteTimestamps"
        static let useBitcoinSymbol = "useBitcoinSymbol"
        static let showFiatBalance = "showFiatBalance"
        static let bitcoinPriceCurrency = "bitcoinPriceCurrency"
        static let checkPendingOnStartup = "checkPendingOnStartup"
        static let checkSentTokens = "checkSentTokens"
        static let autoPasteEcashReceive = "autoPasteEcashReceive"
        static let useWebsockets = "useWebsockets"
        static let enablePaymentRequests = "enablePaymentRequests"
        static let receivePaymentRequestsAutomatically = "receivePaymentRequestsAutomatically"
        static let enableNWC = "enableNWC"
        static let nwcConnections = "nwcConnections"
        static let showP2PKButtonInDrawer = "showP2PKButtonInDrawer"
        static let p2pkKeys = "p2pkKeys"
        static let checkIncomingInvoices = "checkIncomingInvoices"
        static let periodicallyCheckIncomingInvoices = "periodicallyCheckIncomingInvoices"
        static let nostrRelays = "nostrRelays"
        static let nostrSignerType = "nostr_signer_type"
        static let priceEnabled = "priceServiceEnabled"
        static let priceCurrencyCode = "priceServiceCurrencyCode"
        static let cachedBTCPrice = "cachedBTCPrice"
        static let cachedBTCPriceDate = "cachedBTCPriceDate"

        static func cachedBTCPrice(currency: String) -> String {
            "cachedBTCPrice.\(currency.uppercased())"
        }

        static func cachedBTCPriceDate(currency: String) -> String {
            "cachedBTCPriceDate.\(currency.uppercased())"
        }
    }
    
    // NPC
    static let npcEnabled = "npc.enabled"
    static let npcAutomaticClaim = "npc.automaticClaim"
    static let npcSelectedMint = "npc.selectedMint"
    static let npcLastCheck = "npc.lastCheck"
    
    // Price
    static let priceEnabled = "price.enabled"
    static let priceCurrencyCode = "price.currencyCode"
    static let cachedBTCPrice = "price.cachedBTC"
    static let cachedBTCPriceDate = "price.cachedBTCDate"

    static func cachedBTCPrice(currency: String) -> String {
        "price.cachedBTC.\(currency.uppercased())"
    }

    static func cachedBTCPriceDate(currency: String) -> String {
        "price.cachedBTCDate.\(currency.uppercased())"
    }

    static let walletDataKeys = [
        mints,
        activeMintUrl,
        pendingTokens,
        pendingReceiveTokens,
        claimedTokens,
        transactions,
        savedTokens,
        paymentPreimages,
        meltQuoteFees,
        mintQuoteTimestamps,
        processedNPCQuotes
    ]

    static let walletDataLegacyKeys = [
        Legacy.mints,
        Legacy.pendingTokens,
        Legacy.pendingReceiveTokens,
        Legacy.claimedTokens,
        Legacy.savedTokens,
        Legacy.paymentPreimages,
        Legacy.mintQuoteTimestamps
    ]

    static let walletScopedSettingsKeys = [
        enablePaymentRequests,
        receivePaymentRequestsAutomatically,
        enableNWC,
        nwcConnections,
        showP2PKButtonInDrawer,
        p2pkKeys,
        nostrSignerType,
        npcEnabled,
        npcAutomaticClaim,
        npcSelectedMint,
        npcLastCheck
    ]

    static let walletScopedSettingsLegacyKeys = [
        Legacy.enablePaymentRequests,
        Legacy.receivePaymentRequestsAutomatically,
        Legacy.enableNWC,
        Legacy.nwcConnections,
        Legacy.showP2PKButtonInDrawer,
        Legacy.p2pkKeys,
        Legacy.nostrSignerType
    ]

    static var walletBoundaryKeys: [String] {
        walletDataKeys + walletDataLegacyKeys + walletScopedSettingsKeys + walletScopedSettingsLegacyKeys
    }
    
    // Keychain (Secure Storage)
    enum Secure {
        static let mnemonic = "wallet_mnemonic"
        static let nostrPrivateKey = "nostr_private_key"
    }
}

File: android/app/src/main/java/com/cashu/me/Core/Protocols/StorageProtocol.kt
package com.cashu.me.Core.Protocols

interface KeyValueStore {
    fun string(key: String): String?
    fun putString(key: String, value: String?)
    fun remove(key: String)
    fun removePrefix(prefix: String)
}

interface SecureStorage {
    fun loadString(key: String): String?
    fun saveString(key: String, value: String)
    fun delete(key: String)
    fun contains(key: String): Boolean
}

fun SecureStorage.saveMnemonic(mnemonic: String) {
    saveString(StorageKeys.secureWalletMnemonic, mnemonic)
}

fun SecureStorage.loadMnemonic(): String? = loadString(StorageKeys.secureWalletMnemonic)

fun SecureStorage.deleteMnemonic() {
    delete(StorageKeys.secureWalletMnemonic)
}

fun SecureStorage.hasMnemonic(): Boolean = contains(StorageKeys.secureWalletMnemonic)

fun SecureStorage.saveNostrPrivateKey(privateKeyHex: String) {
    saveString(StorageKeys.secureNostrPrivateKey, privateKeyHex)
}

fun SecureStorage.loadNostrPrivateKey(): String? = loadString(StorageKeys.secureNostrPrivateKey)

fun SecureStorage.deleteNostrPrivateKey() {
    delete(StorageKeys.secureNostrPrivateKey)
}

fun SecureStorage.hasNostrPrivateKey(): Boolean = contains(StorageKeys.secureNostrPrivateKey)

object StorageKeys {
    const val walletDataPrefix = "wallet."
    const val settingsDataPrefix = "settings."
    const val npcDataPrefix = "npc."
    const val priceDataPrefix = "price."

    const val walletMints = "wallet.mints"
    const val walletActiveMintUrl = "wallet.activeMintUrl"
    const val walletPendingTokens = "wallet.pendingTokens"
    const val walletPendingReceiveTokens = "wallet.pendingReceiveTokens"
    const val walletClaimedTokens = "wallet.claimedTokens"
    const val walletTransactions = "wallet.transactions"
    const val walletSavedTokens = "wallet.savedTokens"
    const val walletPaymentPreimages = "wallet.paymentPreimages"
    const val walletMeltQuoteFees = "wallet.meltQuoteFees"
    const val walletMintQuoteTimestamps = "wallet.mintQuoteTimestamps"
    const val walletProcessedNPCQuotes = "wallet.processedNPCQuotes"
    const val walletProcessedCashuRequests = "wallet.processedCashuRequests"
    const val cashuRequests = "cashuRequests.v1"
    const val cashuRequestsCurrentId = "cashuRequests.currentId.v1"

    const val settingsUseBitcoinSymbol = "settings.useBitcoinSymbol"
    const val settingsShowFiatBalance = "settings.showFiatBalance"
    const val settingsBitcoinPriceCurrency = "settings.bitcoinPriceCurrency"
    const val settingsCheckPendingOnStartup = "settings.checkPendingOnStartup"
    const val settingsCheckSentTokens = "settings.checkSentTokens"
    const val settingsAutoPasteEcashReceive = "settings.autoPasteEcashReceive"
    const val settingsUseWebsockets = "settings.useWebsockets"
    const val settingsEnablePaymentRequests = "settings.enablePaymentRequests"
    const val settingsReceivePaymentRequestsAutomatically = "settings.receivePaymentRequestsAutomatically"
    const val settingsShowP2PKButtonInDrawer = "settings.showP2PKButtonInDrawer"
    const val settingsP2PKKeys = "settings.p2pkKeys"
    const val settingsCheckIncomingInvoices = "settings.checkIncomingInvoices"
    const val settingsPeriodicallyCheckIncomingInvoices = "settings.periodicallyCheckIncomingInvoices"
    const val settingsNostrRelays = "settings.nostrRelays"
    const val settingsNostrSignerType = "settings.nostrSignerType"
    const val settingsNostrMintBackupEnabled = "settings.nostrMintBackupEnabled"
    const val walletNostrMintBackupLastBackupDate = "wallet.nostrMintBackup.lastBackupDate"
    const val settingsAmountDisplayPrimary = "settings.amountDisplayPrimary"
    const val settingsHomeBalanceUnit = "settings.homeBalanceUnit"
    const val settingsSentryEnabled = "settings.sentryEnabled"
    const val settingsAppLockEnabled = "settings.appLockEnabled"

    const val npcEnabled = "npc.enabled"
    const val npcAutomaticClaim = "npc.automaticClaim"
    const val npcSelectedMint = "npc.selectedMint"
    const val npcLastCheck = "npc.lastCheck"

    const val priceEnabled = "price.enabled"
    const val priceCurrencyCode = "price.currencyCode"
    const val priceCachedBTC = "price.cachedBTC"
    const val priceCachedBTCDate = "price.cachedBTCDate"

    fun priceCachedBTC(currency: String) = "$priceCachedBTC.${currency.uppercase()}"
    fun priceCachedBTCDate(currency: String) = "$priceCachedBTCDate.${currency.uppercase()}"

    const val secureWalletMnemonic = "wallet_mnemonic"
    const val secureNostrPrivateKey = "nostr_private_key"

    val walletBoundaryKeys = setOf(
        walletMints,
        walletActiveMintUrl,
        walletPendingTokens,
        walletPendingReceiveTokens,
        walletClaimedTokens,
        walletTransactions,
        walletSavedTokens,
        walletPaymentPreimages,
        walletMeltQuoteFees,
        walletMintQuoteTimestamps,
        walletProcessedNPCQuotes,
        walletProcessedCashuRequests,
        cashuRequests,
        cashuRequestsCurrentId,
        settingsP2PKKeys,
        npcEnabled,
        npcAutomaticClaim,
        npcSelectedMint,
        npcLastCheck,
    )
}
